### DISCUSSION TITLE
WebSocket not disconnecting properly

### DISCUSSION THREAD

**Discussion by RainzDev** [fastapi/fastapi#11231] [Questions] [answered]
### First Check

- [x] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, WebSocket, WebSocketDisconnect

import json

app = FastAPI()

class ConnectionManager:
    def __init__(self):
	self.active_connections: list[WebSocket] = []

    async def connect(self, websocket: WebSocket):
        await websocket.accept()
	self.active_connections.append(websocket)
	print(self.active_connections)

    def disconnect(self, websocket: WebSocket):
        self.active_connections.remove(websocket)
	print(self.active_connections)

    async def broadcast(self, message: str):
	for connection in self.active_connections:
	    await connection.send_json(message)

manager = ConnectionManager()

@app.websocket('/ws')
async def websocket_endpoint(websocket: WebSocket):
    await manager.connect(websocket)
    try:
	while True:
	    data = await websocket.receive_text()
	    data_js = json.loads(data)
	    await manager.broadcast({"test": "test"})
    except WebSocketDisconnect:
	manager.disconnect(websocket)
```

### Description

Hello! I got some issues with the websocket. For some reason, it doesn't disconnect properly when a connection is closed. I'm getting this error `RuntimeError: Unexpected ASGI message 'websocket.send' after sending 'websocket.close'.` since one or more connections doesnt exist in the self.active_connections list anymore.

i can easily tell that it didnt disconnect properly because these amount of websocket "connections" arent supposed to be in the list since I only connected one

[Here's the print for the whole active_connections list](https://media.discordapp.net/attachments/772631203689922562/1213233108158124103/image.png?ex=65f4ba61&is=65e24561&hm=545b6b05db7250f083436b19ed94916a81b8103d59830b59de0d329902810d21&=&format=webp&quality=lossless&width=598&height=306)

Keep in mind that there are multiple connections in there because I'm using a Discord bot where it uses the lib `websocket-client` to send the text. 

**Expected to happen**

There shouldn't be any of those errors and no connections that aren't meant to exist still after disconnecting

**What actually happened**

Multiple non-existent connections shown in the list.

### Operating System

Linux

### FastAPI Version

0.104.1

### Pydantic Version

2.5.2

### Python Version

Python 3.10.12

**Comment by RainzDev** [comment, marked as answer]
All good now. The issue was because of other errors that wasn't being thrown out.

**Comment by Jaza** [comment]
I experienced the same issue, websockets that appear to be disconnected but to have never resulted in `WebSocketDisconnect` getting raised, so they're hanging around in the `active_connections` list. I resolved it by adding some code to `broadcast()` to detect and clean up such websockets, hope this helps others:

```python
WEBSOCKET_DEAD_ERROR_MESSAGE = (
    "Unexpected ASGI message 'websocket.send', "
    "after sending 'websocket.close' or response already completed."
)


class ConnectionManager:

    # ...

    async def broadcast(self, message: Any):
        dead_connections: list[WebSocket] = []

        for connection in self.active_connections:
            try:
                await connection.send_json(message)
            except RuntimeError as ex:
                if str(ex) == WEBSOCKET_DEAD_ERROR_MESSAGE:
                    logger.warning(
                        "Removing a websocket from active_connections list because "
                        "it appears to be dead, when trying to send to it the "
                        f"following was raised: {WEBSOCKET_DEAD_ERROR_MESSAGE}"
                    )
                    dead_connections.append(connection)
                else:
                    raise ex

        for connection in dead_connections:
            self.disconnect(connection)
```
